File: php/src/ThaiBreak.php
<?php

declare(strict_types=1);

namespace ThaiBreak;

/**
 * ThaiBreak — Primary entry point for Thai word tokenization and typographic line breaking.
 *
 * Fast, high-accuracy Thai text processing for PHP and Laravel.
 *
 * Usage:
 *   use ThaiBreak\ThaiBreak;
 *
 *   // 1. Word Tokenization
 *   $words = ThaiBreak::words('ฉันรักภาษาไทย');
 *   // → ['ฉัน', 'รัก', 'ภาษา', 'ไทย']
 *
 *   // 2. Insert Line Breaks (ZWSP) for dompdf, mPDF, or Web
 *   $pdfSafeHtml = ThaiBreak::lines('<p>บริษัทได้จัดสรรงบประมาณ ฿50,000,000</p>', isHtml: true);
 *
 *   // 3. Soft Wrapping for fixed-width text
 *   $wrapped = ThaiBreak::wrap($longText, width: 45);
 *
 * License: Apache-2.0
 */
class ThaiBreak
{
    /**
     * Tokenize Thai text into words.
     *
     * @param  string $text           UTF-8 text
     * @param  bool   $keepWhitespace Whether to include whitespace tokens
     * @return string[]
     */
    public static function words(string $text, bool $keepWhitespace = false): array
    {
        return ThaiTokenizer::getDefault()->tokenize($text, $keepWhitespace);
    }

    /**
     * Alias of words() — tokenize Thai text into words.
     *
     * @param  string $text           UTF-8 text
     * @param  bool   $keepWhitespace Whether to include whitespace tokens
     * @return string[]
     */
    public static function tokenize(string $text, bool $keepWhitespace = false): array
    {
        return self::words($text, $keepWhitespace);
    }

    /**
     * Tokenize Thai text and join tokens with a delimiter.
     *
     * @param  string $text      UTF-8 text
     * @param  string $delimiter Delimiter string (default '|')
     * @return string
     */
    public static function join(string $text, string $delimiter = '|'): string
    {
        return ThaiTokenizer::getDefault()->tokenizeToString($text, $delimiter);
    }

    /**
     * Insert line break opportunities (Zero-Width Space U+200B by default)
     * adhering to Thai typographic rules (no mid-word cuts, no orphan punctuation).
     *
     * @param  string $text   Plain text or HTML markup
     * @param  string $marker Marker to insert (default: "\u{200B}")
     * @param  bool   $isHtml Whether to preserve HTML tags and entities
     * @return string
     */
    public static function lines(string $text, string $marker = ThaiLineBreaker::DEFAULT_MARKER, bool $isHtml = false): string
    {
        return ThaiTokenizer::getDefault()->insertLineBreaks($text, $marker, $isHtml);
    }

    /**
     * Alias of lines() — insert line break opportunities.
     */
    public static function breakLines(string $text, string $marker = ThaiLineBreaker::DEFAULT_MARKER, bool $isHtml = false): string
    {
        return self::lines($text, $marker, $isHtml);
    }

    /**
     * Soft wrap Thai text to a specified visual column width.
     * Accurately accounts for Thai combining vowels and tone marks (0 width).
     *
     * @param  string $text         UTF-8 plain text
     * @param  int    $width        Maximum column width per line
     * @param  string $break        Line break separator (default "\n")
     * @param  bool   $cutLongWords Whether to force break words wider than $width
     * @return string
     */
    public static function wrap(string $text, int $width = 60, string $break = "\n", bool $cutLongWords = false): string
    {
        return ThaiTokenizer::getDefault()->wrap($text, $width, $break, $cutLongWords);
    }

    /**
     * Calculate visual display column width for Thai text (ignoring combining marks).
     */
    public static function displayWidth(string $text): int
    {
        return ThaiLineBreaker::thaiDisplayWidth($text);
    }

    /**
     * Add custom words with optional weights to the shared tokenizer instance.
     *
     * @param array<string, float>|string[] $words
     */
    public static function addCustomWords(array $words): void
    {
        ThaiTokenizer::getDefault()->addCustomWords($words);
    }

    /**
     * Add a single custom word to the shared tokenizer instance.
     */
    public static function addWord(string $word, float $weight = 1.0): void
    {
        ThaiTokenizer::getDefault()->addWord($word, $weight);
    }

    /**
     * Get the underlying ThaiTokenizer singleton.
     */
    public static function getTokenizer(): ThaiTokenizer
    {
        return ThaiTokenizer::getDefault();
    }

    /**
     * Get the underlying ThaiLineBreaker bound to the default tokenizer.
     */
    public static function getLineBreaker(): ThaiLineBreaker
    {
        return ThaiTokenizer::getDefault()->getLineBreaker();
    }
}

File: go/thaibreak.go
package thaibreak

import (
	"os"
	"path/filepath"
	"strings"
	"sync"
)

var (
	defaultMu          sync.RWMutex
	defaultTokenizer   *Tokenizer
	defaultLineBreaker *LineBreaker
	defaultInitOnce    sync.Once
)

func initDefault() {
	defaultInitOnce.Do(func() {
		// Look for dictionary in common locations
		candidates := []string{
			"../data/words.txt",
			"data/words.txt",
			"../../data/words.txt",
		}
		var dictPath, bigramPath string
		for _, c := range candidates {
			if _, err := os.Stat(c); err == nil {
				dictPath = c
				dir := filepath.Dir(c)
				b := filepath.Join(dir, "bigrams.tsv")
				if _, err2 := os.Stat(b); err2 == nil {
					bigramPath = b
				}
				break
			}
		}

		trie := NewThaiTrie()
		if dictPath != "" {
			if t, err := LoadTsvFile(dictPath); err == nil {
				trie = t
			}
		}

		var bigrams *BigramModel
		if bigramPath != "" {
			if b, err := LoadBigramsFile(bigramPath); err == nil {
				bigrams = b
			}
		}

		defaultTokenizer = NewTokenizer(trie, bigrams)
		defaultLineBreaker = NewLineBreaker(defaultTokenizer)
	})
}

// SetDefault configures the global shared tokenizer instance.
func SetDefault(tok *Tokenizer) {
	defaultMu.Lock()
	defer defaultMu.Unlock()
	defaultTokenizer = tok
	defaultLineBreaker = NewLineBreaker(tok)
}

// GetDefaultTokenizer returns the global shared Tokenizer.
func GetDefaultTokenizer() *Tokenizer {
	initDefault()
	defaultMu.RLock()
	defer defaultMu.RUnlock()
	return defaultTokenizer
}

// GetDefaultLineBreaker returns the global shared LineBreaker.
func GetDefaultLineBreaker() *LineBreaker {
	initDefault()
	defaultMu.RLock()
	defer defaultMu.RUnlock()
	return defaultLineBreaker
}

// Words tokenizes Thai text into words.
func Words(text string) []string {
	return GetDefaultTokenizer().Tokenize(text, false)
}

// Tokenize is an alias of Words.
func Tokenize(text string) []string {
	return Words(text)
}

// Join tokenizes and joins tokens with a delimiter.
func Join(text string, sep string) string {
	return strings.Join(Words(text), sep)
}


// Lines inserts Zero-Width Space (U+200B) break markers adhering to Thai typographic rules.
func Lines(text string, isHtml bool) string {
	return GetDefaultLineBreaker().InsertLineBreaks(text, DefaultBreakMarker, isHtml)
}

// BreakLines is an alias of Lines.
func BreakLines(text string, isHtml bool) string {
	return Lines(text, isHtml)
}

// Wrap soft-wraps Thai text to fit within `width` visual columns.
func Wrap(text string, width int) string {
	return GetDefaultLineBreaker().Wrap(text, width, "\n", false)
}

// DisplayWidth returns the visual column width for Thai text.
func DisplayWidth(text string) int {
	return ThaiDisplayWidth(text)
}
